src/vm_pages.rs

```diff
@@ -471,26 +470,16 @@ impl<'a, T: GuestStagePageTable> ActiveVmPages<'a, T> {
         F: FnMut(GuestPhysAddr, *const u8, usize) -> usize,
     {
         let this_cpu = PerCpu::this_cpu();
-        let mut copied = 0;
-        let mut cur_gpa = guest_addr;
-        let mut cur_ptr = host_ptr;
-        while copied < len {
-            this_cpu.enter_guest_memcpy();
-            let bytes = copy_fn(cur_gpa, cur_ptr, len - copied);
-            this_cpu.exit_guest_memcpy();
-            copied += bytes;
-            if copied < len {
-                // Partial copy: we encountered a page fault. See if we can handle it and retry.
-                cur_gpa = cur_gpa
-                    .checked_increment(bytes as u64)
-                    .ok_or(Error::AddressOverflow)?;
-                self.vm_pages.handle_page_fault(cur_gpa)?;
-
-                // Safety: cur_ptr + bytes must be less than the original host_ptr + len.
-                cur_ptr = unsafe { cur_ptr.add(bytes) };
-            }
+        this_cpu.enter_guest_memcpy();
+        let bytes = copy_fn(guest_addr, host_ptr, len);
+        this_cpu.exit_guest_memcpy();
+        if bytes == len {
+            Ok(())
+        } else {
+            // TODO: Report faulting address and region so that the caller can turn it into a VM
+            // exit code.
+            Err(Error::UnhandledPageFault)
```

Comment:
What will the flow look like if the guest asks to copy a measurement to an as-yet-unmapped page?

TVM -> ecall (get measuement) -> Fault -> Host(Exit code indicating fault and address) -> Fill page dance -> TvmRun -> Result gets at ecall instruction and restart from scratch?

Reply:
In this case, we'll think that we'll fault in the copy_to_guest(), detect that this a demand-zero page that needs to be mapped in, exit to the host with the address and guest fault address, wait for the map page dance, then following TvmRun, will attempt to copy again. It might get interesting if the host retries without actually mapping a page (in which case we'll have a loop).

Reply:
Yup, exactly that flow. The host is free to continuously try running a vCPU without fixing the page fault if it wants to; denial-of-service by the host is out-of-scope.